ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PmtTransactionPool.java

```diff
@@ -0,0 +1,113 @@
+package org.hyperledger.besu.ethereum.privacy;
+
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.ethereum.chain.BlockAddedEvent;
+import org.hyperledger.besu.ethereum.chain.BlockAddedObserver;
+
+import java.util.HashMap;
+import java.util.Map;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class PmtTransactionPool implements BlockAddedObserver {
+  private static final Logger LOG = LogManager.getLogger();
+  private final Map<Hash, PmtTransactionTracker> pmtPool;
+
+  public PmtTransactionPool() {
+    this.pmtPool = new HashMap<>();
+  }
+
+  @Override
+  public void onBlockAdded(final BlockAddedEvent event) {
+    // remove from the map any transactions that went into this block
+    // TODO take into account re-orgs
+    event
+        .getAddedTransactions()
+        .forEach(
+            tx -> {
+              LOG.debug("removing " + pmtPool.containsKey(tx.getHash()));
+              pmtPool.remove(tx.getHash());
+            });
+  }
+
+  public long getMaxMatchingNonce(final String sender, final String privacyGroupId) {
+
+    return pmtPool.values().stream()
+        .filter(
+            pmt ->
+                pmt.getSender().equals(sender)
+                    && pmt.getPrivacyGroupIdBase64().equals(privacyGroupId))
```

Comment:
Just for me to know: is the private nonce unique per sender address, or per (sender address, privacyGroup)? From the code I assume the later, just want to double check.

Reply:
per (sender and privacyGroup)